### DISCUSSION TITLE
How to monitor multiple Java processes

### DISCUSSION THREAD

**Discussion by pizixi** [ccfos/nightingale#1265] [Q&A]
目前使用 categraf 的 procstat 插件 search_cmdline_substring= "****"
然后 告警规则里面写procstat_lookup_count == 0，这样只能对单个java进程进行监控，多个的话有啥方案？
前期使用open-falcon 直接监控java进程数的方式实现，并且linux和window均可以，夜莺可有类似的方式
![image](https://user-images.githubusercontent.com/56673215/201290713-778f95e1-ab16-48cf-a65e-7a6e0ffcc024.png)

**Comment by FengZh61** [comment]
#1221

**Reply by pizixi** [reply]
那个我看了，我说的是多个java进程

**Reply by FengZh61** [reply]
> 那个我看了，我说的是多个java进程

[[instances]]是个数组，支持多个吧，测试一下

**Reply by pizixi** [reply]
另外其实这种方式，感觉没有我说的open-falcon 方便，因为要到各个服务器上去改配置文件，不是直接在web端设置

**Reply by FengZh61** [reply]
> 另外其实这种方式，感觉没有我说的open-falcon 方便，因为要到各个服务器上去改配置文件，不是直接在web端设置

你可以找个插件远程控制categraf 也能实现统一管理。categraf 目前还是读取的本地配置，没法统一的管理。

**Reply by pizixi** [reply]
请问您知道[[instances]] 这个数组如何配置吗？另外我不是要远程控制categraf， 我看视频上老版本有个采集配置 功能，https://www.bilibili.com/video/BV1dQ4y1N7Aq/?spm_id_from=333.999.0.0，但是新版本没有了， 如果有个类似功能应该就能实现我说的，无需在多个服务器来回改配置文件，然后重启categraf

**Reply by FengZh61** [reply]
> 请问您知道[[instances]] 这个数组如何配置吗？另外我不是要远程控制categraf， 我看视频上老版本有个采集配置 功能，[https://www.bilibili.com/video/BV1dQ4y1N7Aq/?spm_id_from=333.999.0.0，但是新版本没有了，](https://www.bilibili.com/video/BV1dQ4y1N7Aq/?spm_id_from=333.999.0.0%EF%BC%8C%E4%BD%86%E6%98%AF%E6%96%B0%E7%89%88%E6%9C%AC%E6%B2%A1%E6%9C%89%E4%BA%86%EF%BC%8C) 如果有个类似功能应该就能实现我说的，无需在多个服务器来回改配置文件，然后重启categraf

[[instances]]
search_cmdline_substring = "abc.properties"
gather_total = true
gather_per_pid = false
gather_more_metrics = [
    "threads",
    "fd",
    "io",
    "uptime",
    "cpu",
    "mem",
    "limit"
]
[[instances]]
search_cmdline_substring = "def.properties"
gather_total = true
gather_per_pid = false
gather_more_metrics = [
    "threads",
    "fd",
    "io",
    "uptime",
    "cpu",
    "mem",
    "limit"
]

**Reply by pizixi** [reply]
测试过了，无效的

**Reply by FengZh61** [reply]
> 测试过了，无效的

![image](https://user-images.githubusercontent.com/98309003/201314134-18d268b4-5e3f-4383-be3e-662779cd5a56.png)

![image](https://user-images.githubusercontent.com/98309003/201314314-8acde28c-a1f3-413e-8aad-ef693995c1d8.png)


我的测试结果，我理解这是支持多个进程的。

**Reply by pizixi** [reply]
谢谢，我再试试

**Reply by pizixi** [reply]
另外，请问我新建告警规则的时候 如何选择单个服务器，我看目前只能选择 生效集群  ，有很多服务器上的进程是不一样的，都是独立特性，这样不会只能 每台服务器分一个 业务组吧，然后再在每个业务组上 新增一个进程告警规则 ？

**Reply by FengZh61** [reply]
> 另外，请问我新建告警规则的时候 如何选择单个服务器，我看目前只能选择 生效集群 ，有很多服务器上的进程是不一样的，都是独立特性，这样不会只能 每台服务器分一个 业务组吧，然后再在每个业务组上 新增一个进程告警规则 ？

每个服务器都有自己唯一的信息，比如ip、hostname、id等，你可以根据这些信息在label筛选出指定的服务器，如果现在的metric中没有这些信息，你可以加上。

**Reply by pizixi** [reply]
非常感谢，我在PromQL 中加上了ident ，完美解决了
